rust/lance/src/dataset/blob.rs

```diff
@@ -0,0 +1,318 @@
+        .await
+    }
+
+    /// Read up to `len` bytes from the current cursor position
+    ///
+    /// After this call the cursor will be pointing to the end of
+    /// the read data.
+    pub async fn read_up_to(&self, len: usize) -> Result<bytes::Bytes> {
+        let position = self.position;
+        let size = self.size;
+        self.do_with_reader(|cursor, reader| async move {
+            let start = position as usize + cursor as usize;
+            let size = len.min((size - cursor) as usize);
+            let end = position as usize + size;
+            let data = reader.get_range(start..end).await?;
+            Ok((end as u64, data))
+        })
+        .await
+    }
+
+    /// Seek to a new cursor position in the file
+    pub async fn seek(&self, new_cursor: u64) -> Result<()> {
+        let mut reader = self.reader.lock().await;
+        match reader.deref_mut() {
+            ReaderState::Open((cursor, _)) => {
+                *cursor = new_cursor;
+                Ok(())
+            }
+            ReaderState::Closed => Err(Error::IO {
+                location: location!(),
+                source: "Blob file is already closed".into(),
+            }),
+            ReaderState::Uninitialized(cursor) => {
+                *cursor = new_cursor;
+                Ok(())
+            }
+        }
+    }
+
+    /// Return the current cursor position in the file
+    pub async fn tell(&self) -> Result<u64> {
+        let reader = self.reader.lock().await;
+        match *reader {
+            ReaderState::Open((cursor, _)) => Ok(cursor),
+            ReaderState::Closed => Err(Error::IO {
+                location: location!(),
+                source: "Blob file is already closed".into(),
+            }),
+            ReaderState::Uninitialized(cursor) => Ok(cursor),
+        }
+    }
+
+    /// Return the size of the blob file in bytes
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+}
+
+pub(super) async fn take_blobs(
+    dataset: &Arc<Dataset>,
+    row_ids: &[u64],
+    column: &str,
+) -> Result<Vec<BlobFile>> {
+    let projection = dataset.schema().project(&[column])?;
+    let blob_field = &projection.fields[0];
+    let blob_field_id = blob_field.id;
+    if blob_field.data_type() != DataType::LargeBinary
+        || !projection.fields[0].metadata.contains_key(BLOB_META_KEY)
+    {
+        return Err(Error::InvalidInput {
+            location: location!(),
+            source: format!("the column '{}' is not a blob column", column).into(),
+        });
+    }
+    let description_and_addr = dataset
+        .take_builder(row_ids, projection)?
+        .with_row_address(true)
+        .execute()
+        .await?;
+    let descriptions = description_and_addr.column(0).as_struct();
+    let positions = descriptions.column(0).as_primitive::<UInt64Type>();
+    let sizes = descriptions.column(1).as_primitive::<UInt64Type>();
+    let row_addrs = description_and_addr.column(1).as_primitive::<UInt64Type>();
```

Comment:
Should we add an assert that they have the same number of rows? I forget whether `take` guarantees that. Consider the case where the row_id has been deleted.

Reply:
Ah, good idea

Reply:
I added a test and added the assert.  For now I just raise a NotImplementedError.

The current `_take_rows` behavior is to filter out deleted items but I'm not sure that's ideal.  It is impossible for the caller to know which IDs were deleted which seems like it would be useful information.

We can either change the behavior to "fill with NULL" or we can make it configurable in a future PR.